sequencer/src/state.rs

```diff
@@ -1,47 +1,146 @@
+    block::{BlockMerkleTree, FeeAccount, FeeAmount, FeeMerkleTree},
+    Header, Payload,
+};
 use commit::{Commitment, Committable};
-use ethers::prelude::BlockNumber;
 use hotshot::traits::State as HotShotState;
-use hotshot_types::data::ViewNumber;
+use hotshot_types::data::{BlockError, ViewNumber};
+use jf_primitives::merkle_tree::{
+    AppendableMerkleTreeScheme, MerkleTreeScheme, UniversalMerkleTreeScheme,
+};
 use serde::{Deserialize, Serialize};
-use std::fmt::Debug;
+use tracing::debug;
 
-#[derive(Default, Serialize, Deserialize, Clone, Debug, Hash, PartialEq, Eq)]
-pub struct State {
-    chain_variables: ChainVariables,
-    ethereum_block_tag: BlockNumber,
+#[derive(Hash, Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct ValidatedState {
+    /// Frontier of Block Merkle Tree
+    pub block_merkle_tree: BlockMerkleTree,
+    /// Fee Merkle Tree
+    pub fee_merkle_tree: FeeMerkleTree,
 }
 
-impl HotShotState for State {
-    type Error = Error;
+impl Default for ValidatedState {
+    fn default() -> Self {
+        let block_merkle_tree =
+            BlockMerkleTree::from_elems(32, Vec::<Commitment<Header>>::new()).unwrap();
+
+        // Words of wisdom from @mrain: "capacity = arity^height"
+        // "For index space 2^160, arity 256 (2^8),
+        // you should set the height as 160/8=20"
+        let fee_merkle_tree =
+            FeeMerkleTree::from_kv_set(20, Vec::<(FeeAccount, FeeAmount)>::new()).unwrap();
+        Self {
+            block_merkle_tree,
+            fee_merkle_tree,
+        }
+    }
+}
+
+impl ValidatedState {
+    fn validate_proposal(&self, parent: &Header, proposal: &Header) -> anyhow::Result<Self> {
+        // validate height
+        anyhow::ensure!(
+            proposal.height == parent.height + 1,
+            anyhow::anyhow!(
+                "Invalid Height Error: {}, {}",
+                parent.height,
+                proposal.height
+            )
+        );
+
+        // validate parent fee merkle tree root against state
+        let fee_merkle_tree = self.fee_merkle_tree;
+        let fee_merkle_tree_root = fee_merkle_tree.commitment();
+        anyhow::ensure!(
+            parent.fee_merkle_tree_root == fee_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Fee Merkle Tree Error: {}, {}",
+                parent.fee_merkle_tree_root,
+                fee_merkle_tree_root
+            )
+        );
+
+        // validate parent block merkle tree root against state
+        // `clone` to avoid possible side-effects
+        let mut block_merkle_tree = self.block_merkle_tree.clone();
+        let block_merkle_tree_root = block_merkle_tree.commitment();
+        anyhow::ensure!(
+            parent.block_merkle_tree_root == block_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Block Merkle Tree Error: {}, {}",
+                parent.block_merkle_tree_root,
+                block_merkle_tree_root
+            )
+        );
```

Comment:
I don't think we need to validate this, though. We are keeping track of `parent` and the corresponding state (`self`) locally, so we should be able to trust that the correspond